contracts/contracts/Tally.sol

```diff
@@ -0,0 +1,248 @@
+import {VkRegistry} from "./VkRegistry.sol";
+
+
+contract Tally is
+    Ownable,
+    SnarkCommon,
+    Hasher
+{
+    // Error codes
+    string constant ERROR_VOTING_PERIOD_NOT_PASSED = "TallyE01";
+    string constant ERROR_PROCESSING_NOT_COMPLETE = "TallyE02";
+    string constant ERROR_INVALID_TALLY_VOTES_PROOF = "TallyE03";
+    string constant ERROR_ALL_BALLOTS_TALLIED = "TallyE04";
+
+    uint8 private constant LEAVES_PER_NODE = 5;
+
+    // The commitment to the tally results. Its initial value is 0, but after
+    // the tally of each batch is proven on-chain via a zk-SNARK, it should be
+    // updated to:
+    //
+    // hash3(
+    //   hashLeftRight(merkle root of current results, salt0)
+    //   hashLeftRight(number of spent voice credits, salt1),
+    //   hashLeftRight(merkle root of the no. of spent voice credits per vote option, salt2)
+    // )
+    //
+    // Where each salt is unique and the merkle roots are of arrays of leaves
+    // TREE_ARITY ** voteOptionTreeDepth long.
+    uint256 public tallyCommitment;
+
+    uint256 public tallyBatchNum;
+
+    // The final commitment to the state and ballot roots
+    uint256 public sbCommitment;
+
+    Verifier public verifier;
+
+    constructor(Verifier _verifier) {
+        verifier = _verifier;
+    }
+
+    modifier votingPeriodOver(Poll _poll) {
+        (uint256 deployTime, uint256 duration) = _poll
+            .getDeployTimeAndDuration();
+        // Require that the voting period is over
+        uint256 secondsPassed = block.timestamp - deployTime;
+        require(secondsPassed > duration, ERROR_VOTING_PERIOD_NOT_PASSED);
+        _;
+    }
+
+    /*
+     * Pack the batch start index and number of signups into a 100-bit value.
+     */
+    function genTallyVotesPackedVals(
+        uint256 _numSignUps,
+        uint256 _batchStartIndex,
+        uint256 _tallyBatchSize
+    ) public pure returns (uint256) {
+        // TODO: ensure that each value is less than or equal to 2 ** 50
+        uint256 result = (_batchStartIndex / _tallyBatchSize) +
+            (_numSignUps << uint256(50));
+
+        return result;
+    }
+
+    function genTallyVotesPublicInputHash(
+        uint256 _numSignUps,
+        uint256 _batchStartIndex,
+        uint256 _tallyBatchSize,
+        uint256 _newTallyCommitment
+    ) public view returns (uint256) {
+        uint256 packedVals = genTallyVotesPackedVals(
+            _numSignUps,
+            _batchStartIndex,
+            _tallyBatchSize
+        );
+        uint256[] memory input = new uint256[](4);
+        input[0] = packedVals;
+        input[1] = sbCommitment;
+        input[2] = tallyCommitment;
+        input[3] = _newTallyCommitment;
+        uint256 inputHash = sha256Hash(input);
+        return inputHash;
+    }
+
+    // TODO: make sure correct mp address is passed or change to private function
+    // TODO: reuse tally.sol for multiple polls
+    function updateSbCommitment(MessageProcessor _mp) public {
```

Comment:
let's make this private for now? it's only called in tallyVotes and tallyVotes is owner only so should be ok

Reply:
it has been called in proveOnChain.ts. The reason is that after the deployment of the new Tally.sol contract. We have to update the sbCommitment for the first time by calling this function. So if we change it to private, we cannot call it. Let me rephrase the TODO here.